llmfoundry/data/packing.py

```diff
@@ -2,29 +2,70 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import os
-from typing import Any, Callable, Dict, List, Literal, Optional, Tuple
+from typing import Any, Callable, Dict, Iterable, List, Literal, Optional, Tuple
 
 import numpy as np
 import torch
 from omegaconf import DictConfig
 from transformers import PreTrainedTokenizerBase
-
-
-class BinPackWrapper:
+from torch.utils.data import IterableDataset
+
+
+class BinPackDataset(IterableDataset):
+    """An IterableDataset that returns packed examples."""
+    def __init__(
+        self,
+        dataset: IterableDataset,
+        packing_ratio: int,
+        target_batch_size: int,
+        max_seq_len: int,
+        pad_token_id: int,
+    ):
+        self.dataset = dataset
+        self.packing_ratio = int(packing_ratio)
+        self.out_size = int(target_batch_size)
+        self.max_seq_len = int(max_seq_len)
+        self.pad_token_id = int(pad_token_id)
+        self.collator = BinPackCollator(
+            lambda x: x, 
+            target_batch_size=self.out_size, 
+            max_seq_len=max_seq_len, 
+            pad_token_id=pad_token_id, 
+            max_leftover_bins_to_keep=None # Keep all leftovers.
+        )
+    def __iter__(self) -> Iterable:
+        examples = []
+        for example in self.dataset:
+            examples.append(example)
+            if len(examples) == self.packing_ratio * self.out_size:
+                packed_examples = self.collator(examples)
+                print('len packed examples', len(packed_examples))
+                for packed_example in packed_examples:
+                    yield packed_example
+                examples = []
+        # Finish the last batch
+        packed_examples = self.collator(examples)
+        for packed_example in packed_examples:
+            yield packed_example
+        examples = []
+
+        # Iterate over leftovers.
+        for _, leftover in self.collator._leftover_bins:
+            yield leftover
+
+class BinPackCollator:
```

Comment:
todo: remove this class altogether in favor of BinPackDataset, logic from __call__ should be moved to BinPackDataset class